### DISCUSSION TITLE
How to load an audio file with a custom sample rate - 1024 Hz?

### DISCUSSION THREAD

**Discussion by dizcza** [katspaugh/wavesurfer.js#2768] [Q&A]
I'm reading a WAV file from an SD card of my ESP32 board serving a web file server.

The official wavesurfer examples work fine but the problem is that my audio file sample rate is 1024 Hz. This is not an audio file, in fact, it's the output of an I2C differential pressure sensor that outputs at a 1024 Hz sample rate.

I've been searching for a solution to (1) load a custom raw audio file and (2) plot its spectrogram. WaveSurfer AudioContext does not support such a sample rate:

```
Uncaught DOMException: AudioContext constructor: Sample rate 1024 is not in the range [8000, 192000]
```

The full script:

```html
<!DOCTYPE html>
<head><meta charset='UTF-8'/>
<script src="https://unpkg.com/wavesurfer.js"></script>
</head>

<body>
<div id="waveform"></div>
<div id="wave-spectrogram"></div>

<button class="btn btn-primary" onclick="wavesurfer.playPause()">
  <i class="glyphicon glyphicon-play"></i>
  Play/Pause
</button>
</body>

<script>
const audioCtx = new AudioContext( { sampleRate: 1024 } );

var wavesurfer = WaveSurfer.create({
    container: '#waveform',
    waveColor: '#D2EDD4',
    progressColor: '#46B54D',
    audioContext: audioCtx
});

wavesurfer.on('ready', function () {
  var spectrogram = Object.create(WaveSurfer.Spectrogram);
  spectrogram.init({
    wavesurfer: wavesurfer,
    container: "#wave-spectrogram",
    fftSamples: 128,
    labels: true
  });
});

wavesurfer.load('/fs/SDP_amplified.wav');
</script>
</html>
```

Is it a strong NO for my case and should I look for something else?

Thanks.

**Comment by dizcza** [comment]
This audio file can be easily imported into Audacity, for example, where I can specify the input sample rate. And I can play it successfully.

**Comment by katspaugh** [comment]
The lowest sample rate in Web Audio is indeed 8kHz, although I found Chrome able to go down to 3kHz.
It should however resample a lower/higher sample rate audio file to its own sample rate. What happens if you don’t specify an AudioContext?

**Comment by dizcza** [comment]
> What happens if you don’t specify an AudioContext?

Nothing happens. The file is not loaded and I cannot play it.

**Reply by katspaugh** [reply]
Have you tried with the alpha version?

**Reply by dizcza** [reply]
> Have you tried with the alpha version?

Unfortunately, I cannot install node wavesurfer.js on my ESP32 board - it simply does not have enough flash space (we're talking about a total of 300 Kb available).

I tried installing it on my Linux desktop machine as suggested in the alpha version

```
npm install --save wavesurfer.js@alpha
```

and then navigated to the `node_modules/wavesurfer.js/dist` and put `app.js` with the following content:

```node
import WaveSurfer from 'wavesurfer.js'

const wavesurfer = WaveSurfer.create({
  container: document.body,
  waveColor: 'rgb(200, 0, 200)',
  progressColor: 'rgb(100, 0, 100)',
  url: 'http://192.168.3.92/fs/demo.wav',
})

wavesurfer.once('interaction', () => {
  wavesurfer.play()
})
```

The URL points to the ESP32 file server (I can download this file). When I try to launch it with `node app.js`, it throws me the error:

```
file:///home/dizcza/tools/js/wavesurfer.ts/node_modules/wavesurfer.js/dist/wavesurfer.js:42
        const url = this.options.url || this.options.media?.src;
                                                           ^

SyntaxError: Unexpected token '.'
    at Loader.moduleStrategy (internal/modules/esm/translators.js:133:18)
```

I ain't much familiar with node js. Have I had an option of installing it on my ESP32, I'd have looked into this issue. But I'm looking for a solution that can do everything on the client side, i.e. downloading a script in the HTML header that does the job.

**Reply by katspaugh** [reply]
It's a browser program, you shouldn't run it on node.

You can use a script tag like in your first code snippet:
```
<script src="https://unpkg.com/wavesurfer.js@alpha"></script>
```

**Reply by dizcza** [reply]
Ah OK, I tried the alpha and got

1. if I don't specify `audioContext`:
```
caught (in promise) DOMException: Failed to execute 'decodeAudioData' on 'BaseAudioContext': Unable to decode audio data
```

2. if I specify `audioContext`:
```
caught DOMException: Failed to construct 'AudioContext': The hardware sample rate provided (1024) is outside the range [3000, 768000].
```

Here is the link to a sample file recorded at 1024 Hz that you can try: https://www.dropbox.com/s/qk9j81uz5suwuzx/sample1024hz.wav?dl=1. The recording plays wind for about 10 sec.

**Reply by dizcza** [reply]
There were some errors like cannot import from outside of a module so I tested with the following script:

```html
<!DOCTYPE html>
<head><meta charset='UTF-8'/>
<script type="module" src="https://unpkg.com/wavesurfer.js@alpha">
</script>
</head>

<body>
</body>

<script type="module">

import WaveSurfer from 'https://unpkg.com/wavesurfer.js@alpha'

const audioCtx = new AudioContext( { sampleRate: 1024 } );

const wavesurfer = WaveSurfer.create({
  container: document.body,
  waveColor: 'rgb(200, 0, 200)',
  progressColor: 'rgb(100, 0, 100)',
  audioContext: audioCtx,
  url: '/fs/SDP_amplified.wav',
})

wavesurfer.once('interaction', () => {
  wavesurfer.play()
})
</script>

</html>
```

**Reply by katspaugh** [reply]
Like I already said, browsers don't support sample rates below 8k as per Web Audio specs (although Chrome/Safari support down to 3000 Hz).

The error `Unable to decode audio data` is also from Web Audio. The browser cannot decode your file.

You can either pre-decode the audio using a tool like [audiowaveform](https://github.com/bbc/audiowaveform) and pass that data to wavesurfer in the `peaks` option, or [resample the wav](https://stackoverflow.com/questions/67880409/ffmpeg-how-to-resample-audio-file) to 8kHz.

If you go the pre-decoded peaks route:
```
<!DOCTYPE html>
<head><meta charset='UTF-8'/>
</head>

<body>
</body>

<script type="module">

import WaveSurfer from 'https://unpkg.com/wavesurfer.js@alpha'

const wavesurfer = WaveSurfer.create({
  container: document.body,
  waveColor: 'rgb(200, 0, 200)',
  progressColor: 'rgb(100, 0, 100)',
  url: '/fs/SDP_amplified.wav',
  peaks: [[ 0, 0.1, 0.2 ... 0.8, 0.5, 0 ]]
})

wavesurfer.once('interaction', () => {
  wavesurfer.play()
})
</script>

</html>
```

**Reply by dizcza** [reply]
Yeah, I was thinking about upsampling. Is there an approach in wavesurfer to
1. fetch a raw audio file (WAV) into a buffer (either in chunks or completely)
2. upsample the buffer (copy each int16_t sample N times)
3. feed the upsampled buffer into the playback
4. feed the upsampled buffer into the spectrogram
5. display on-the-fly or once fully completed

?

**Reply by katspaugh** [reply]
Playback isn't a problem. It's the decoding that fails. wavesurfer needs to decode the file to render it.
So you don't need to resample it in JS, you only need to parse it, strip out the wav headers and get the raw PCM data. See [this lib](https://github.com/chdh/wav-file-decoder/blob/main/src/WavFileDecoder.ts) for inspiration.
Once you get the PCM data, just use the code snippet I posted above to pass it as the `peaks` option.

**Reply by dizcza** [reply]
In fact, I'm writing the audio file in raw PCM data, and I mentioned WAV only because it's a more understandable format.

I was able to convert my raw int16_t audio data into peaks with the lib you suggested and the plot is rendered correctly. However, it does not play automatically. How can I plot the waveform, play it and show the spectrogram? Since your alpha version does not support spectrogram plots so far, I switched back to the stable release.

Here is the updated code that includes the spectrogram part:

```html
<!DOCTYPE html>
<head><meta charset='UTF-8'/>
<script src="https://unpkg.com/wavesurfer.js"></script>
</head>

<body>
<div id="waveform"></div>
<div id="wave-spectrogram"></div>

<button class="btn btn-primary" onclick="wavesurfer.playPause()">
  <i class="glyphicon glyphicon-play"></i>
  Play/Pause
</button>
</body>

<script>

var wavesurfer = WaveSurfer.create({
    container: '#waveform',
    waveColor: '#D2EDD4',
    progressColor: '#46B54D',
    normalize: true,
    plugins: [
        WaveSurfer.spectrogram.create({
            container: "#wave-spectrogram",
            labels: true,
            height: 256,
        })
    ]
});

wavesurfer.load('/fs/SDP_amplified.wav', [a list of unnormalized 8-bit pixes]);
</script>
</html>
```

This brings me

```
caught TypeError: Cannot read properties of undefined (reading 'create')
```

This is unexpected as I basically follow the [official example](https://wavesurfer-js.org/example/spectrogram/index.html).

And if I use [this](https://codepen.io/entonbiba/pen/Bpmqrq) example, no spectrogram is shown (although it seems to work on codepen.io with the demo.wav file).

Furthermore, I also don't know where I should specify something for the `preload` and/or `duration` options of the `load()` function.

**Reply by katspaugh** [reply]
You need to import also the spectrogram plugin script

**Reply by dizcza** [reply]
> You need to import also the spectrogram plugin script

True! I haven't thought about this as my intuition was that everything is in the wavesurfer.js file. It's also not mentioned on the docs ([here](https://wavesurfer-js.org/plugins/spectrogram.html) and [here](https://wavesurfer-js.org/example/spectrogram/index.html); it's only said on the main Plugins page that I hadn't visited).

So only the "how to play a 1024 sample rate raw audio file" question is left, the other two are resolved. Any suggestions?

----

Update 1

Actually, no spectrogram is shown for this 1024 sample rate file when I specify the peaks. Only the waveform is visible. But it does show a spectrogram of the official demo.wav file.

Update 2

When I compute peaks of the demo.wav manually, the spectrogram is also not shown. It's displayed as soon as I click the play button and is not visible till that. So I think resolving the playing issue will resolve the spectrogram not showing on the page.

**Reply by katspaugh** [reply]
Looks like the spectrogram code expects the file to be decoded by Web Audio and ignores the peaks.
To solve the playback issue you can use `backend: “MediaElement”`.
Spectrogram needs code changes to use peaks.

**Reply by dizcza** [reply]
When I set the backend to "MediaElement" and click Play, I get
```
wavesurfer.js:2017 Uncaught (in promise) DOMException: The element has no supported sources.
play	@	wavesurfer.js:2017
play	@	wavesurfer.js:4146
playPause	@	wavesurfer.js:4184
onclick	@	waveform:11
```
and it does not play

**Reply by katspaugh** [reply]
The Spectrogram plugin is now available also in the alpha version, and there it will use pre-decoded peaks if provided.
Still your main problem is probably playback, which apparently browsers don't support.

**Reply by katspaugh** [reply]
Someone has asked a similar question recently: https://github.com/wavesurfer-js/wavesurfer.js/discussions/2774#discussioncomment-5815931
Maybe that solution would also work for you. Basically, load a "fake" audio URL, but pass the correct pre-decoded peaks and duration.

**Reply by dizcza** [reply]
Yeah, I come back to the upsampling idea again: upsampling 3X will solve all three issues (Chrome is able to open an audio file with the 3k rate). Copying the element 3 times introduces spurious frequency in the upsampled audio signal. So I was thinking to 
1) open a stream buffer in JS (read a file in chunks or read it completely)
2) apply the 3x upsample FIR filter (or do it in the FFT - overkill)
3) feed the stream to WaveSurfer playback, waveform, and spectrogram calculation.

I'm not sure if WaveSurfer is able to do such things, i.e. providing a `File` object instead of the URL. In any case, I haven't seen any JS code to do upsampling without introducing spurious frequencies.

**Reply by katspaugh** [reply]
You can always [make a URL out of a file/blob](https://developer.mozilla.org/en-US/docs/Web/API/URL/createObjectURL_static).
I can't advise anything regarding upsampling though. Curious to know if you get it working!
